File: app/routers/estoque.py
from fastapi import APIRouter, Depends, HTTPException, Request
from fastapi.templating import Jinja2Templates
from sqlalchemy.orm import Session
from typing import List, Optional
from decimal import Decimal

from app.database import get_db
from app.models import Produto, MovimentoEstoque
from app.services.estoque_service import EstoqueService

router = APIRouter()
templates = Jinja2Templates(directory="app/templates")

@router.get("/estoque")
async def tela_estoque(request: Request):
    """Tela de controle de estoque"""
    return templates.TemplateResponse(
        "estoque.html", 
        {"request": request}
    )

@router.get("/api/estoque/alertas")
async def alertas_estoque(db: Session = Depends(get_db)):
    """Obter alertas de estoque baixo e alto"""
    estoque_service = EstoqueService(db)
    
    produtos_estoque_baixo = estoque_service.verificar_estoque_baixo()
    produtos_estoque_alto = estoque_service.verificar_estoque_alto()
    
    return {
        "estoque_baixo": [
            {
                "id": p.id,
                "nome": p.nome,
                "codigo_barras": p.codigo_barras,
                "estoque_atual": p.estoque,
                "estoque_minimo": p.estoque_minimo,
                "categoria": p.categoria
            }
            for p in produtos_estoque_baixo
        ],
        "estoque_alto": [
            {
                "id": p.id,
                "nome": p.nome,
                "codigo_barras": p.codigo_barras,
                "estoque_atual": p.estoque,
                "estoque_maximo": p.estoque_maximo,
                "categoria": p.categoria
            }
            for p in produtos_estoque_alto
        ]
    }

@router.post("/api/estoque/entrada")
async def entrada_estoque(
    produto_id: int,
    quantidade: int,
    preco_unitario: Optional[float] = None,
    motivo: str = "Entrada de estoque",
    documento: Optional[str] = None,
    db: Session = Depends(get_db)
):
    """Registrar entrada de estoque"""
    try:
        estoque_service = EstoqueService(db)
        movimento = estoque_service.entrada_estoque(
            produto_id=produto_id,
            quantidade=quantidade,
            preco_unitario=Decimal(str(preco_unitario)) if preco_unitario else None,
            motivo=motivo,
            documento=documento
        )
        
        return {
            "id": movimento.id,
            "produto_id": movimento.produto_id,
            "tipo_movimento": movimento.tipo_movimento,
            "quantidade": movimento.quantidade,
            "quantidade_anterior": movimento.quantidade_anterior,
            "quantidade_nova": movimento.quantidade_nova,
            "motivo": movimento.motivo,
            "created_at": movimento.created_at.isoformat()
        }
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))

@router.post("/api/estoque/ajuste")
async def ajuste_estoque(
    produto_id: int,
    quantidade_nova: int,
    motivo: str = "Ajuste de estoque",
    db: Session = Depends(get_db)
):
    """Ajustar estoque para quantidade específica"""
    try:
        estoque_service = EstoqueService(db)
        movimento = estoque_service.ajuste_estoque(
            produto_id=produto_id,
            quantidade_nova=quantidade_nova,
            motivo=motivo
        )
        
        return {
            "id": movimento.id,
            "produto_id": movimento.produto_id,
            "tipo_movimento": movimento.tipo_movimento,
            "quantidade": movimento.quantidade,
            "quantidade_anterior": movimento.quantidade_anterior,
            "quantidade_nova": movimento.quantidade_nova,
            "motivo": movimento.motivo,
            "created_at": movimento.created_at.isoformat()
        }
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))

@router.get("/api/estoque/movimentos")
async def listar_movimentos(
    produto_id: Optional[int] = None,
    tipo_movimento: Optional[str] = None,
    limit: int = 100,
    db: Session = Depends(get_db)
):
    """Listar movimentações de estoque"""
    estoque_service = EstoqueService(db)
    movimentos = estoque_service.obter_movimentos(
        produto_id=produto_id,
        tipo_movimento=tipo_movimento,
        limit=limit
    )
    
    return [
        {
            "id": m.id,
            "produto_id": m.produto_id,
            "produto_nome": m.produto.nome,
            "tipo_movimento": m.tipo_movimento,
            "quantidade": m.quantidade,
            "quantidade_anterior": m.quantidade_anterior,
            "quantidade_nova": m.quantidade_nova,
            "preco_unitario": float(m.preco_unitario) if m.preco_unitario else None,
            "valor_total": float(m.valor_total) if m.valor_total else None,
            "motivo": m.motivo,
            "observacoes": m.observacoes,
            "usuario": m.usuario,
            "documento": m.documento,
            "created_at": m.created_at.isoformat()
        }
        for m in movimentos
    ]

@router.get("/api/estoque/relatorio")
async def relatorio_estoque(db: Session = Depends(get_db)):
    """Relatório geral de estoque"""
    produtos = db.query(Produto).filter(Produto.ativo == True).all()
    
    total_produtos = len(produtos)
    total_valor_estoque = sum(float(p.preco) * p.estoque for p in produtos)
    
    # Contar por categoria
    categorias = {}
    for produto in produtos:
        cat = produto.categoria or "Sem categoria"
        if cat not in categorias:
            categorias[cat] = {"quantidade": 0, "valor": 0}
        categorias[cat]["quantidade"] += produto.estoque
        categorias[cat]["valor"] += float(produto.preco) * produto.estoque
    
    # Produtos sem movimento
    estoque_service = EstoqueService(db)
    produtos_estoque_baixo = len(estoque_service.verificar_estoque_baixo())
    produtos_estoque_alto = len(estoque_service.verificar_estoque_alto())
    
    return {
        "resumo": {
            "total_produtos": total_produtos,
            "total_valor_estoque": total_valor_estoque,
            "produtos_estoque_baixo": produtos_estoque_baixo,
            "produtos_estoque_alto": produtos_estoque_alto
        },
        "categorias": categorias,
        "produtos": [
            {
                "id": p.id,
                "nome": p.nome,
                "categoria": p.categoria,
                "estoque_atual": p.estoque,
                "estoque_minimo": p.estoque_minimo,
                "estoque_maximo": p.estoque_maximo,
                "preco": float(p.preco),
                "valor_total": float(p.preco) * p.estoque,
                "status": (
                    "baixo" if p.estoque <= p.estoque_minimo else
                    "alto" if p.estoque >= p.estoque_maximo else
                    "normal"
                )
            }
            for p in produtos
        ]
    }

File: app/static/js/estoque.js
// Inventory control functionality
document.addEventListener('DOMContentLoaded', function() {
    let produtos = [];
    let produtosFiltrados = [];
    let movimentacoes = [];
    let currentAjusteProduto = null;

    // Load initial data
    loadAlertas();
    loadRelatorio();
    loadMovimentacoes();

    // Event listeners
    document.getElementById('btnNovaEntrada').addEventListener('click', () => openModal('modalNovaEntrada'));
    document.getElementById('btnInventario').addEventListener('click', () => openModal('modalInventario'));
    document.getElementById('buscaProduto').addEventListener('input', debounce(filtrarProdutos, 300));
    document.getElementById('filtroCategoria').addEventListener('change', filtrarProdutos);
    document.getElementById('filtroStatus').addEventListener('change', filtrarProdutos);

    // Load alerts
    async function loadAlertas() {
        try {
            const response = await apiRequest('/api/estoque/alertas');
            
            // Estoque baixo
            const alertaBaixo = document.getElementById('alertaEstoqueBaixo');
            const countBaixo = document.getElementById('countEstoqueBaixo');
            const listaBaixo = document.getElementById('produtosEstoqueBaixo');
            
            if (response.estoque_baixo.length > 0) {
                alertaBaixo.style.display = 'block';
                countBaixo.textContent = response.estoque_baixo.length;
                listaBaixo.innerHTML = response.estoque_baixo.map(p => `
                    <div class="alert-item">
                        <strong>${p.nome}</strong>
                        <span>Atual: ${p.estoque_atual} | Mín: ${p.estoque_minimo}</span>
                        <button class="btn-mini btn-primary" onclick="abrirAjuste(${p.id})">Ajustar</button>
                    </div>
                `).join('');
            }
            
            // Estoque alto
            const alertaAlto = document.getElementById('alertaEstoqueAlto');
            const countAlto = document.getElementById('countEstoqueAlto');
            const listaAlto = document.getElementById('produtosEstoqueAlto');
            
            if (response.estoque_alto.length > 0) {
                alertaAlto.style.display = 'block';
                countAlto.textContent = response.estoque_alto.length;
                listaAlto.innerHTML = response.estoque_alto.map(p => `
                    <div class="alert-item">
                        <strong>${p.nome}</strong>
                        <span>Atual: ${p.estoque_atual} | Máx: ${p.estoque_maximo}</span>
                    </div>
                `).join('');
            }
            
        } catch (error) {
            console.error('Erro ao carregar alertas:', error);
            showNotification('Erro ao carregar alertas de estoque', 'error');
        }
    }

    // Load inventory report
    async function loadRelatorio() {
        try {
            const response = await apiRequest('/api/estoque/relatorio');
            
            // Update summary
            document.getElementById('totalProdutos').textContent = response.resumo.total_produtos;
            document.getElementById('valorTotalEstoque').textContent = formatCurrency(response.resumo.total_valor_estoque);
            document.getElementById('produtosEstoqueBaixoCount').textContent = response.resumo.produtos_estoque_baixo;
            
            // Store products
            produtos = response.produtos;
            produtosFiltrados = [...produtos];
            
            // Populate category filter
            const categorias = [...new Set(produtos.map(p => p.categoria).filter(Boolean))];
            const filtroCategoria = document.getElementById('filtroCategoria');
            filtroCategoria.innerHTML = '<option value="">Todas as categorias</option>' +
                categorias.map(cat => `<option value="${cat}">${cat}</option>`).join('');
            
            // Load products for entrada modal
            const entradaProduto = document.getElementById('entradaProduto');
            entradaProduto.innerHTML = '<option value="">Selecione um produto...</option>' +
                produtos.map(p => `<option value="${p.id}">${p.nome} (${p.codigo_barras || 'S/C'})</option>`).join('');
            
            renderProdutos();
            
        } catch (error) {
            console.error('Erro ao carregar relatório:', error);
            showNotification('Erro ao carregar dados do estoque', 'error');
        }
    }

    // Render products table
    function renderProdutos() {
        const tbody = document.querySelector('#tabelaEstoque tbody');
        
        if (produtosFiltrados.length === 0) {
            tbody.innerHTML = '<tr><td colspan="8" class="empty">Nenhum produto encontrado</td></tr>';
            return;
        }
        
        tbody.innerHTML = produtosFiltrados.map(produto => {
            const statusClass = {
                'baixo': 'status-low',
                'alto': 'status-high',
                'normal': 'status-normal'
            }[produto.status];
            
            const statusText = {
                'baixo': 'Baixo',
                'alto': 'Alto',
                'normal': 'Normal'
            }[produto.status];
            
            return `
                <tr>
                    <td>
                        <div class="produto-info">
                            <strong>${produto.nome}</strong>
                            <small>${produto.codigo_barras || 'Sem código'}</small>
                        </div>
                    </td>
                    <td>${produto.categoria || 'Sem categoria'}</td>
                    <td class="text-center"><strong>${produto.estoque_atual}</strong></td>
                    <td class="text-center">${produto.estoque_minimo}/${produto.estoque_maximo}</td>
                    <td>${formatCurrency(produto.preco)}</td>
                    <td>${formatCurrency(produto.valor_total)}</td>
                    <td><span class="status-badge ${statusClass}">${statusText}</span></td>
                    <td class="actions">
                        <button class="btn-mini btn-secondary" onclick="abrirAjuste(${produto.id})" title="Ajustar estoque">
                            ⚖️
                        </button>
                    </td>
                </tr>
            `;
        }).join('');
    }

    // Filter products
    function filtrarProdutos() {
        const busca = document.getElementById('buscaProduto').value.toLowerCase();
        const categoria = document.getElementById('filtroCategoria').value;
        const status = document.getElementById('filtroStatus').value;
        
        produtosFiltrados = produtos.filter(produto => {
            const matchBusca = !busca || 
                produto.nome.toLowerCase().includes(busca) ||
                (produto.codigo_barras && produto.codigo_barras.includes(busca));
            
            const matchCategoria = !categoria || produto.categoria === categoria;
            const matchStatus = !status || produto.status === status;
            
            return matchBusca && matchCategoria && matchStatus;
        });
        
        renderProdutos();
    }

    // Load movements
    async function loadMovimentacoes() {
        try {
            const response = await apiRequest('/api/estoque/movimentos?limit=50');
            movimentacoes = response;
            renderMovimentacoes();
        } catch (error) {
            console.error('Erro ao carregar movimentações:', error);
        }
    }

    // Render movements table
    function renderMovimentacoes() {
        const tbody = document.querySelector('#tabelaMovimentacoes tbody');
        
        if (movimentacoes.length === 0) {
            tbody.innerHTML = '<tr><td colspan="7" class="empty">Nenhuma movimentação encontrada</td></tr>';
            return;
        }
        
        tbody.innerHTML = movimentacoes.map(mov => {
            const tipoClass = {
                'entrada': 'tipo-entrada',
                'saida': 'tipo-saida',
                'venda': 'tipo-venda',
                'ajuste_positivo': 'tipo-ajuste-pos',
                'ajuste_negativo': 'tipo-ajuste-neg'
            }[mov.tipo_movimento] || '';
            
            return `
                <tr>
                    <td>${new Date(mov.created_at).toLocaleDateString('pt-BR')} ${new Date(mov.created_at).toLocaleTimeString('pt-BR', {hour: '2-digit', minute: '2-digit'})}</td>
                    <td>${mov.produto_nome}</td>
                    <td><span class="tipo-badge ${tipoClass}">${mov.tipo_movimento}</span></td>
                    <td class="text-center">${mov.tipo_movimento.includes('negativo') || mov.tipo_movimento === 'venda' || mov.tipo_movimento === 'saida' ? '-' : '+'}${mov.quantidade}</td>
                    <td class="text-center">${mov.quantidade_anterior} → ${mov.quantidade_nova}</td>
                    <td>${mov.motivo || '-'}</td>
                    <td>${mov.usuario}</td>
                </tr>
            `;
        }).join('');
    }

    // Global functions
    window.openModal = function(modalId) {
        document.getElementById(modalId).style.display = 'flex';
    };

    window.closeModal = function(modalId) {
        document.getElementById(modalId).style.display = 'none';
        // Reset forms
        if (modalId === 'modalNovaEntrada') {
            document.getElementById('formNovaEntrada').reset();
        } else if (modalId === 'modalAjusteEstoque') {
            document.getElementById('formAjusteEstoque').reset();
            currentAjusteProduto = null;
        }
    };

    window.abrirAjuste = function(produtoId) {
        const produto = produtos.find(p => p.id === produtoId);
        if (!produto) return;
        
        currentAjusteProduto = produto;
        
        // Fill modal
        document.getElementById('ajusteProdutoInfo').innerHTML = `
            <div class="produto-card">
                <h4>${produto.nome}</h4>
                <p>Código: ${produto.codigo_barras || 'N/A'}</p>
                <p>Categoria: ${produto.categoria || 'Sem categoria'}</p>
            </div>
        `;
        
        document.getElementById('ajusteQuantidadeAtual').textContent = produto.estoque_atual;
        document.getElementById('ajusteQuantidadeNova').value = produto.estoque_atual;
        
        openModal('modalAjusteEstoque');
    };

    window.salvarEntrada = async function() {
        const form = document.getElementById('formNovaEntrada');
        const formData = new FormData(form);
        
        const data = {
            produto_id: parseInt(document.getElementById('entradaProduto').value),
            quantidade: parseInt(document.getElementById('entradaQuantidade').value),
            preco_unitario: parseFloat(document.getElementById('entradaPreco').value) || null,
            motivo: document.getElementById('entradaMotivo').value,
            documento: document.getElementById('entradaDocumento').value || null
        };
        
        if (!data.produto_id || !data.quantidade) {
            showNotification('Preencha todos os campos obrigatórios', 'error');
            return;
        }
        
        try {
            await apiRequest('/api/estoque/entrada', {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify(data)
            });
            
            showNotification('Entrada registrada com sucesso!', 'success');
            closeModal('modalNovaEntrada');
            
            // Reload data
            loadAlertas();
            loadRelatorio();
            loadMovimentacoes();
            
        } catch (error) {
            console.error('Erro ao salvar entrada:', error);
            showNotification('Erro ao registrar entrada', 'error');
        }
    };

    window.salvarAjuste = async function() {
        if (!currentAjusteProduto) return;
        
        const quantidadeNova = parseInt(document.getElementById('ajusteQuantidadeNova').value);
        const motivo = document.getElementById('ajusteMotivo').value;
        
        if (quantidadeNova < 0) {
            showNotification('Quantidade não pode ser negativa', 'error');
            return;
        }
        
        try {
            await apiRequest('/api/estoque/ajuste', {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({
                    produto_id: currentAjusteProduto.id,
                    quantidade_nova: quantidadeNova,
                    motivo: motivo
                })
            });
            
            showNotification('Ajuste realizado com sucesso!', 'success');
            closeModal('modalAjusteEstoque');
            
            // Reload data
            loadAlertas();
            loadRelatorio();
            loadMovimentacoes();
            
        } catch (error) {
            console.error('Erro ao salvar ajuste:', error);
            showNotification('Erro ao realizar ajuste', 'error');
        }
    };

    window.iniciarInventario = function() {
        // Redirect to inventory audit page
        window.location.href = '/inventario';
    };

    console.log('Inventory control initialized');
});